components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/AppSearchWidgetProvider.kt

```diff
@@ -0,0 +1,282 @@
+        val showMic = (voiceSearchIntent != null)
+        val layout = getLayout(layoutSize, showMic)
+        val text = getText(layoutSize, context)
+
+        val views =
+            createRemoteViews(context, layout, textSearchIntent, voiceSearchIntent, text)
+        appWidgetManager.updateAppWidget(appWidgetId, views)
+    }
+
+    private fun createRemoteViews(
+        context: Context,
+        layout: Int,
+        textSearchIntent: PendingIntent,
+        voiceSearchIntent: PendingIntent?,
+        text: String?
+    ): RemoteViews {
+        return RemoteViews(context.packageName, layout).apply {
+            setIcon(context)
+            when (layout) {
+                R.layout.mozac_search_widget_extra_small_v1,
+                R.layout.mozac_search_widget_extra_small_v2,
+                R.layout.mozac_search_widget_small_no_mic -> {
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab,
+                        textSearchIntent
+                    )
+                }
+                R.layout.mozac_search_widget_small -> {
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab,
+                        textSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_voice,
+                        voiceSearchIntent
+                    )
+                }
+                R.layout.mozac_search_widget_medium,
+                R.layout.mozac_search_widget_large -> {
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab,
+                        textSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_voice,
+                        voiceSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab_icon,
+                        textSearchIntent
+                    )
+                    setTextViewText(R.id.mozac_button_search_widget_new_tab, text)
+                    // Unlike "small" widget, "medium" and "large" sizes do not have separate layouts
+                    // that exclude the microphone icon, which is why we must hide it accordingly here.
+                    if (voiceSearchIntent == null) {
+                        setViewVisibility(R.id.mozac_button_search_widget_voice, View.GONE)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun RemoteViews.setIcon(context: Context) {
+        // gradient color available for android:fillColor only on SDK 24+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            setImageViewResource(
+                R.id.mozac_button_search_widget_new_tab_icon,
+                R.drawable.mozac_search_widget_icon
```

Comment:
For `R.drawable.mozac_search_widget_icon`, since we want the icon to be customizable, you could try to use simple data classes to hold configurations that the app can set. Similar to how you have a `abstract fun` you can have `abstract val`. This way we can land this component without needing to wait for designs/product direction on what the icon/configuration should look like, we can do the final configuration setup in Focus' integration side since we know we roughly want the same widget in Focus/Fenix.

For example:

```kotlin
abstract class AppSearchWidgetProvider {
  abstract val config: SearchWidgetConfig
}

data class SearchWidgetConfig(
  val iconResource: Int
  // insert other properties here that you'd like to make configurable on the app side.
)
```

Reply:
Done